<a href="https://colab.research.google.com/github/RudreshVeerkhare/Data_Science_Lab/blob/main/Hypothesis_Testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Name - **Rudresh Veerkhare**  
UID - **2018130061**

In [1]:
# Donwloading data
!gdown --id "1ifgvz5fUFPsziwMYnr2Lx2TfxlcTFysi" -O data.csv

/usr/local/lib/python3.7/dist-packages/gdown/cli.py:131: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  category=FutureWarning,
Downloading...
From: https://drive.google.com/uc?id=1ifgvz5fUFPsziwMYnr2Lx2TfxlcTFysi
To: /content/data.csv
100% 3.52M/3.52M [00:00<00:00, 193MB/s]


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns; sns.set(style = "whitegrid", rc = {"figure.figsize": (16, 12)})
import warnings
warnings.filterwarnings("ignore")

df_org = pd.read_csv("data.csv")
df_org.head()

age        work_class  final_weight  education  education_num  \
0   39         State-gov         77516  Bachelors             13   
1   50  Self-emp-not-inc         83311  Bachelors             13   
2   38           Private        215646    HS-grad              9   
3   53           Private        234721       11th              7   
4   28           Private        338409  Bachelors             13   

       marital_status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital_gain  capital_loss  hours_per_week native_country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [7]:
# replace question mark with nan
df = df_org.copy()
df[df_org  == '?'] = np.nan

In [8]:
(df.isna().sum() * 100 / len(df)).sort_values(ascending = False)

occupation        5.660146
work_class        5.638647
native_country    1.790486
age               0.000000
final_weight      0.000000
education         0.000000
education_num     0.000000
marital_status    0.000000
relationship      0.000000
race              0.000000
sex               0.000000
capital_gain      0.000000
capital_loss      0.000000
hours_per_week    0.000000
income            0.000000
dtype: float64

Impute missing values, for **occupation**, **work_class** and **native_country**  
All 3 are categorical features so we'll replace it with mode of  respective columns

In [9]:
df["occupation"] = df["occupation"].fillna(df["occupation"].mode()[0])
df["work_class"] = df["work_class"].fillna(df["work_class"].mode()[0])
df["native_country"] = df["native_country"].fillna(df["native_country"].mode()[0])

In [10]:
df["native_country"].mode()[0]

'United-States'

In [11]:
 df.head()

age        work_class  final_weight  education  education_num  \
0   39         State-gov         77516  Bachelors             13   
1   50  Self-emp-not-inc         83311  Bachelors             13   
2   38           Private        215646    HS-grad              9   
3   53           Private        234721       11th              7   
4   28           Private        338409  Bachelors             13   

       marital_status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital_gain  capital_loss  hours_per_week native_country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

# Hypothesis Analysis

My Hypothesis regarding this data is that female income is less than male income.

So, my hypothesis are

$H_0$ - **Female income greater than equal to Male income**   
$H_1$ - **Female income is less than Male income**

But, our data is categorical, i.e Inocome is either high or low we'll caluate probablities of male and female having high income and compair that  

In [12]:
male_income = df["income"][df["sex"] == "Male"].map({'<=50K': "low", '>50K': "high"})
female_income = df["income"][df["sex"] == "Female"].map({'<=50K': "low", '>50K': "high"})

probality_male_high_income = sum(male_income == "high") / len(male_income)
probablity_female_high_income = sum(female_income == "high") / len(female_income)
print(f"Probablity That Female has high income : {probablity_female_high_income}")
print(f"Probablity That Male has high income : {probality_male_high_income}")

Probablity That Female has high income : 0.10946058861758426
Probablity That Male has high income : 0.3057365764111978



$P_{female}$ - Probablity that Female has high income  
$P_{male}$ - Probablity that Male has high income  

So now our Hypothesis would be  

$H_0$ : $P_{female} = P_{male}$    
$H_1$ : $P_{female} \lt P_{male}$


$P_{male}$ : $0.3057365764111978$  
$N_{male}$ : $21,790$

Also take level of signficance to be 0.05  
$α = 0.05$

but also,  
$α:$ Probablity of Type $I$ error  
$α = CDF(Z)$



where,  
$CDF$ - Cumulative Distribution Function with  
$μ = np$  
$σ = \sqrt{np(1-p)}$  
$Z = \frac{X - μ}{σ}$ 

Therefore, 
$X = PPF(α) × σ + μ$  
$PPF$ - Inverse of CDF  

From here we'll get critical value $(X)$

In [13]:
import numpy as np
from scipy.stats import norm

alpha = 0.05
n = len(female_income)
p = probality_male_high_income

mean = n * p
std_dev = (n * p * (1 - p))**0.5

critical_value = norm.ppf(alpha, loc = mean, scale = std_dev) * std_dev + mean

So, if female count having high salary is less than `critical_value` then we successfully reject null hypothsis. 

In [15]:
female_count_with_high_salary = sum(female_income == "high")
print(f"Female count with high salary < critical_value = {female_count_with_high_salary < critical_value}")
print(f"Female count with high salary = {female_count_with_high_salary}")
print(f"critical_value = {critical_value}")

Female count with high salary < critical_value = True
Female count with high salary = 1179
critical_value = 156991.4312984871


This shows that we can successfully reject $H_0$ i.e. $H_1$ is true.  
So, to conclude, probablity that females fall under higher salary range is less than male.